### PR TITLE
Add automated secrets rotation

### PR DESCRIPTION
**Pull request by sethvargo**
**Release Note**

```release-note
**Add automated secrets rotation.** This introduces automated rotation for most application-level secrets. Whereas previously it was the responsibility of server administrators to rotate secrets, the application will now rotate a variety of secrets on regular intervals for improved security.
```

### CODE REVIEW

File: pkg/database/database.go [hunk 1]
```diff
@@ -103,6 +106,132 @@ func (db *Database) KeyManager() keys.KeyManager {
 	return db.keyManager
 }
 
+// GetCookieHashAndEncryptionKeys gets the cookie hash and encryption keys. The
+// first 32 bytes are the encryption key and the remaining bytes are the HMAC
+// key.
+func (db *Database) GetCookieHashAndEncryptionKeys(fallback []envconfig.Base64Bytes) ([][]byte, error) {
+	ctx, done := context.WithTimeout(context.Background(), 5*time.Second)
```
Comment (sethvargo) [inline pkg/database/database.go:113]: Ideally this context would flow through all callers, but that's a fairly large refactor that I didn't want to tackle here. What do y'all think?

File: pkg/database/database.go [hunk 2]
```diff
@@ -103,6 +106,132 @@ func (db *Database) KeyManager() keys.KeyManager {
...
+
+	// TODO(sethvargo): remove in 0.28.0+
+	if len(results) == 0 {
+		existing := envconfigBytesToBytes(fallback)
+		if l := len(existing); l%2 != 0 {
+			return nil, fmt.Errorf("fallback bytes are not even (got %d)", l)
+		}
+
+		// In v1 cookies, the HMAC key is first, then the encryption key.
+		final := make([][]byte, len(existing)/2)
```
Comment (sethvargo) [inline pkg/database/database.go:129]: I hate this.

File: terraform/service_rotation.tf [hunk 3]
```diff
@@ -57,10 +57,17 @@ locals {
   ])
 }
 
+resource "google_secret_manager_secret_iam_member" "rotation-secrets-accessor" {
```
Comment (sethvargo) [inline terraform/service_rotation.tf:60]: I forgot that versionManager doesn't actually include permissions to access the secret. Oops.

File: pkg/database/database.go (hunk 1 above)
Comment (mikehelmick) [reply pkg/database/database.go:113]: I can live w/ this

Comment (mikehelmick) [reply pkg/database/database.go:113]: and refactor later...

File: pkg/database/database.go (hunk 2 above)
Comment (mikehelmick) [reply pkg/database/database.go:129]: it gets removed in the next release at least :)

**Review by mikehelmick — COMMENTED** [review]
LGTM - I'll approve when tests are passing.

Question: How will our local dev environments upgrade / seed in the new world?

Should secrets be added to the seed script?

**Comment by sethvargo** [comment]
> Question: How will our local dev environments upgrade / seed in the new world?

Local dev won't upgrade. Drop and re-create the database. You can remove the associated secrets from your .env file.

> Should secrets be added to the seed script?

They already are - that's the reason the Rotate function is public. It was in the previous PR.

**Comment by sethvargo** [comment]
Don't merge. There's an occasional test failure when running in race mode that I'm looking into...

**Review by mikehelmick — APPROVED** [review]

**Pull request merged**